tempesta_fw/ss_skb.c

```diff
@@ -533,80 +615,74 @@ __split_pgfrag_del(struct sk_buff *skb, int i, int off, int len, TfwStr *it)
+	if (__extend_pgfrags(skb, bskb, i + 1, 1))
 		return -EFAULT;
 
 	/* Find the SKB for tail data. */
-	skb_dst = (i < MAX_SKB_FRAGS - 1) ? skb : it->skb;
+	skb_dst = (i < MAX_SKB_FRAGS - 1) ? skb : __next_skb_frag(skb, bskb);
 
 	/* Calculate the length of the tail part. */
 	tail_len = skb_frag_size(frag) - off - len;
 
-	/* Trim the fragment with the head part. */
-	skb_frag_size_sub(frag, len + tail_len);
-
 	/* Make the fragment with the tail part. */
 	i = (i + 1) % MAX_SKB_FRAGS;
 	__skb_fill_page_desc(skb_dst, i, skb_frag_page(frag),
 			     frag->page_offset + off + len, tail_len);
 	__skb_frag_ref(frag);
 
+	/* Trim the fragment with the head part. */
+	skb_frag_size_sub(frag, len + tail_len);
+
 	/* Adjust SKB data lengths. */
-	ss_skb_adjust_data_len(skb, -len);
 	if (skb != skb_dst) {
-		ss_skb_adjust_data_len(skb, -tail_len);
+		if (skb != bskb)
+			ss_skb_adjust_data_len(skb, -tail_len);
 		ss_skb_adjust_data_len(skb_dst, tail_len);
 	}
+	ss_skb_adjust_data_len(skb, -len);
+	ss_bskb_adjust_len(skb, bskb, -len);
 
-	/* Get the SKB and the address of data after the deleted area. */
-	it->flags = (skb != skb_dst);
+	/* Get the SKB and the address for data after the deleted data. */
 	it->ptr = skb_frag_address(&skb_shinfo(skb_dst)->frags[i]);
+	it->skb = bskb;
 	return 0;
 
 lookup_next_ptr:
-	/* Get the next fragment after the deleted fragment. */
-	if (i < si->nr_frags)
+	/* Get the address of data after the deleted data. */
+	if (i < si->nr_frags) {
 		it->ptr = skb_frag_address(&si->frags[i]);
+		it->skb = bskb;
+	} else {
+		struct sk_buff *dskb;
+		it->skb = __next_skb(skb, &bskb, &dskb);
+		it->ptr = __skb_data_address(dskb, bskb);
+	}
 	return 0;
 }
 
 static int
-__split_pgfrag(struct sk_buff *skb, int i, int off, int len, TfwStr *it)
+__split_pgfrag(struct sk_buff *skb, struct sk_buff *bskb,
+	       int i, int off, int len, TfwStr *it)
 {
 	return len > 0
-		? __split_pgfrag_add(skb, i, off, len, it)
-		: __split_pgfrag_del(skb, i, off, -len, it);
-}
-
-/*
- * Determine the address of data in @skb.
- *
- * Note that @skb is not expected to have
- * SKB fragments without page fragments.
- */
-static inline void *
-ss_skb_data_address(struct sk_buff *skb)
-{
-	if (skb == NULL)
-		return NULL;
-	if (skb_headlen(skb))
-		return skb->data;
-	BUG_ON(!skb_is_nonlinear(skb));
-	BUG_ON(!skb_shinfo(skb)->nr_frags);
-	return skb_frag_address(&skb_shinfo(skb)->frags[0]);
+		? __split_pgfrag_add(skb, bskb, i, off, len, it)
+		: __split_pgfrag_del(skb, bskb, i, off, -len, it);
 }
 
 /**
- * Fragment @skb to add some room if @len > 0 or delete data otherwise.
+ * Add room for data to @skb if @len > 0 or delete data otherwise.
+ * Most of the time that is done by fragmenting the @skb.
+ * @bskb is the base skb if @skb is an skb fragment.
+ * Otherwise it equals @skb.
  */
 static int
-__skb_fragment(struct sk_buff *skb, struct sk_buff *pskb,
+__skb_fragment(struct sk_buff *skb, struct sk_buff *bskb,
```

Comment:
Well comment https://github.com/natsys/tempesta/pull/458#discussion_r61856771 is still not fixed. You've just renamed `pskb` to `bskb`, but the only one place where `bskb` differs from `skb` is internal `__skb_fragment()` call. So the function should be split: `skb_fragment()` should accept `skb` only while `__skb_fragment()`, called from `skb_fragment()` should accept `bskb` as well as `skb`.